fq_test.go

```diff
@@ -0,0 +1,176 @@
+package fq
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"runtime"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+// adapted from https://github.com/tadglines/wfq/blob/master/wfq_test.go
+
+type flowDesc struct {
+	// In
+	ftotal int     // Total units in flow
+	imin   float64 // Min Packet servicetime
+	imax   float64 // Max Packet servicetime
+
+	// Out
+	idealPercent  float64
+	actualPercent float64
+}
+
+func genFlow(fq *FQScheduler, desc *flowDesc, key int) {
+	for i, t := 1, 0; t < desc.ftotal; i++ {
+		it := new(Packet)
+		it.queueidx = key
+		if desc.imin == desc.imax {
+			it.servicetime = desc.imax
+		} else {
+			it.servicetime = min + rand.Float64()*(desc.imax-desc.imin)
+		}
+		if t+it.servicetime > desc.ftotal {
+			it.servicetime = desc.ftotal - t
+		}
+		t += it.servicetime
+		it.seq = i
+		// new packet
+		fq.Enqueue(it)
+	}
+}
+
+func consumeQueue(t *testing.T, fq *FQScheduler, descs []flowDesc) (float64, error) {
+	active := make(map[int]bool)
+	var total int
+	acnt := make(map[int]int)
+	cnt := make(map[int]int)
+	seqs := make(map[int]int)
+
+	wsum := uint64(len(descs))
+
+	for i, ok := fq.Dequeue(); ok; i, ok = fq.Dequeue() {
+		// callback to update virtualtime w/ correct service time for request
+		fq.FinishPacket(i)
+
+		it := i
+		seq := seqs[it.queueidx]
+		if seq+1 != it.seq {
+			return 0, fmt.Errorf("Packet for flow %d came out of queue out-of-order: expected %d, got %d", it.queueidx, seq+1, it.seq)
+		}
+		seqs[it.queueidx] = it.seq
+
+		// set the flow this item is a part of to active
+		active[it.queueidx] = true
+
+		cnt[it.queueidx] += it.servicetime
```

Comment:
Mismatched types here.  Might as well make the `cnt` and `acnt` maps to `float64` too.

Reply:
This has been resolved.